File: clients/vscode/test/materialSnapshotClient.test.js
"use strict";
const test = require("node:test"), assert = require("node:assert/strict");
const client = require("../materialSnapshotClient");
const query = require("../../../core/tests/fixtures/check-snapshot-v1/query.json");
const response = require("../../../core/tests/fixtures/check-snapshot-v1/response.json");

test("historical capture outcome and current reader support remain separate in the summary", () => {
  const { materialSummary } = require("../materialChecksClient");
  const text = materialSummary({ format: client.VIEW, state: "completed", native_outcome: "native-failed", coverage: "complete",
    interpretation: { state: "unsupported", unsupported_sections: ["crafting-values"] },
    overview_state: "loaded", detail_state: "not-loaded", findings: [], findings_count: 7 });
  assert.match(text, /Initialization: native-failed; observation coverage: complete/);
  assert.match(text, /Reader interpretation: unsupported; unsupported sections: crafting-values/);
});

test("snapshot pages agree with the independent Core contract fixture", () => {
  assert.equal(client.validatePage(response, query), response);
  assert.equal(client.identity("fixture", { view_id: "vue 🌍 é", cursor: null, offset: 2 }),
    "fixture:sha256:089c2f2c49da75b273674f7755397eacc6525b42c6d5e43be8c0fd248ec88482");
});
test("snapshot page identity refuses another view, snapshot and page", () => {
  for (const patch of [{ snapshot_id: "check-snapshot:sha256:" + "1".repeat(64) }, { view_id: "new-view" },
    { cursor: { snapshot_id: query.snapshot_id, query_id: response.query_id, offset: 1 } }])
    assert.throws(() => client.validatePage(response, { ...query, ...patch }), /selected view or page/);
  for (const mutate of [v => { v.payload.offset = 1; }, v => { v.payload.total = 2; },
    v => { v.complete = false; }, v => { v.next_cursor = { offset: 2 }; }]) {
    const altered = structuredClone(response); mutate(altered);
    assert.throws(() => client.validatePage(altered, query), /count|progress|cursor/);
  }
});
test("unavailable and unsupported evidence never appear as an empty completed page", () => {
  for (const state of ["unavailable", "unsupported", "expired", "cancelled", "incomplete"]) {
    const value = { ...response, state, payload: null, complete: false, reason: "Explicit fixture reason" };
    assert.equal(client.validatePage(value, query), value);
    assert.throws(() => client.validatePage({ ...value, complete: true }, query), /complete data/);
  }
});
test("native diagnostic reads bind their side and original record address", () => {
  const view = { snapshot_id: query.snapshot_id, view_id: query.view_id };
  const read = client.diagnosticQuery(view, { side: "candidate", pointer: "/result/candidate/result/execution/diagnostics/7" });
  assert.equal(read.section_id, "candidate-diagnostics"); assert.equal(read.record_key, "item:7");
  assert.throws(() => client.diagnosticQuery(view, { side: "baseline", pointer: "/result/candidate/result/execution/diagnostics/7" }), /pointer/);
});

test("bulk paging changes query identity while retaining the exact next record offset", () => {
  const view = { snapshot_id: query.snapshot_id, view_id: query.view_id, finding_page: { next_cursor: { offset: 73 } } };
  const bulk = client.nextFindingsQuery(view, true), small = client.nextFindingsQuery(view, false);
  assert.equal(bulk.preferred_bytes, 1048576); assert.equal(small.preferred_bytes, 65536);
  assert.equal(bulk.cursor.offset, 73); assert.notEqual(bulk.cursor.query_id, small.cursor.query_id);
  const { cursor, ...base } = bulk; assert.equal(cursor.query_id, client.identity("check-snapshot-query", base));
});

File: clients/intellij-community/src/test/java/dev/cleanroommc/workbench/intellij/community/MaterialSnapshotClientTest.java
package dev.cleanroommc.workbench.intellij.community;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class MaterialSnapshotClientTest {
    @Test void retentionTransportBindsExactProposalAndStructuredSettings() {
        String session = "work-session-v2-" + "a".repeat(32);
        String confirmation = "check-retention-proposal:sha256:" + "b".repeat(64);
        var options = JsonParser.parseString("{\"operation\":\"configure\",\"settings\":{\"mode\":\"finite\",\"known_stores\":[\"/path with spaces\"]}}").getAsJsonObject();
        var args = MaterialChecksClient.arguments(session, "retention", null, confirmation, options);
        assertEquals(List.of("--settings", options.get("settings").toString(), "--confirm", confirmation), args.subList(args.size()-4, args.size()));
        assertThrows(IllegalArgumentException.class, () -> MaterialChecksClient.arguments(session, "retention", null, "yes", options));
        options.addProperty("operation", "purge-all");
        assertThrows(IllegalArgumentException.class, () -> MaterialChecksClient.arguments(session, "retention", null, null, options));
    }
    @Test void readerSupportCannotRewriteHistoricalCaptureOrNativeOutcome() {
        var view = JsonParser.parseString("""
            {"format":"workbench-material-check-view-v1","state":"completed",
             "native_outcome":"native-failed","coverage":"complete","findings_count":7,
             "interpretation":{"state":"unsupported","unsupported_sections":["crafting-values"]}}
            """).getAsJsonObject();
        var text = MaterialChecksClient.summary(view);
        assertTrue(text.contains("Initialization: native-failed; observation coverage: complete"));
        assertTrue(text.contains("Reader interpretation: unsupported"));
        assertTrue(text.contains("crafting-values"));
    }
    private JsonObject fixture(String name) throws Exception {
        return JsonParser.parseString(Files.readString(Path.of("../../core/tests/fixtures/check-snapshot-v1/" + name + ".json"))).getAsJsonObject();
    }
    @Test void coreContractFixtureAndUnicodeIdentityAgree() throws Exception {
        var query = fixture("query"); var response = fixture("response");
        assertSame(response, MaterialSnapshotClient.validatePage(response, query));
        var unicode = JsonParser.parseString("{\"view_id\":\"vue 🌍 é\",\"cursor\":null,\"offset\":2}").getAsJsonObject();
        assertEquals("fixture:sha256:089c2f2c49da75b273674f7755397eacc6525b42c6d5e43be8c0fd248ec88482", MaterialSnapshotClient.identity("fixture", unicode));
    }
    @Test void oldViewAndOutOfRangePageCannotReplaceCurrentEvidence() throws Exception {
        var query = fixture("query"); var response = fixture("response");
        query.addProperty("view_id", "another-view");
        assertThrows(IllegalArgumentException.class, () -> MaterialSnapshotClient.validatePage(response, query));
        var original = fixture("query"); response.getAsJsonObject("payload").addProperty("total", 2);
        assertThrows(IllegalArgumentException.class, () -> MaterialSnapshotClient.validatePage(response, original));
    }
    @Test void unavailableStatesNeverCertifyEmptySuccess() throws Exception {
        var query = fixture("query");
        for (String state : List.of("unavailable", "unsupported", "expired", "cancelled", "incomplete")) {
            var response = fixture("response"); response.addProperty("state", state); response.add("payload", com.google.gson.JsonNull.INSTANCE);
            response.addProperty("reason", "Explicit fixture reason"); response.addProperty("complete", false);
            assertSame(response, MaterialSnapshotClient.validatePage(response, query)); response.addProperty("complete", true);
            assertThrows(IllegalArgumentException.class, () -> MaterialSnapshotClient.validatePage(response, query));
        }
    }
    @Test void diagnosticAddressRemainsBoundToSelectedSide() {
        var view = new JsonObject(); view.addProperty("snapshot_id", "fixture"); view.addProperty("view_id", "fixture-view");
        var finding = JsonParser.parseString("{\"side\":\"candidate\",\"pointer\":\"/result/candidate/result/execution/diagnostics/7\"}").getAsJsonObject();
        var query = MaterialSnapshotClient.diagnosticQuery(view, finding);
        assertEquals("candidate-diagnostics", query.get("section_id").getAsString());
        assertEquals("item:7", query.get("record_key").getAsString());
        finding.addProperty("side", "baseline"); assertThrows(IllegalArgumentException.class, () -> MaterialSnapshotClient.diagnosticQuery(view, finding));
    }
    @Test void bulkPagesChangeQueryIdentityWithoutChangingTheRecordOffset() throws Exception {
        var view = fixture("query");
        view.add("finding_page", JsonParser.parseString("{\"next_cursor\":{\"offset\":73}}"));
        var bulk = MaterialSnapshotClient.nextFindingsQuery(view, true);
        var small = MaterialSnapshotClient.nextFindingsQuery(view, false);
        assertEquals(1048576, bulk.get("preferred_bytes").getAsInt());
        assertEquals(65536, small.get("preferred_bytes").getAsInt());
        assertEquals(73, bulk.getAsJsonObject("cursor").get("offset").getAsInt());
        assertNotEquals(bulk.getAsJsonObject("cursor").get("query_id"), small.getAsJsonObject("cursor").get("query_id"));
        var base = bulk.deepCopy(); base.remove("cursor");
        assertEquals(MaterialSnapshotClient.identity("check-snapshot-query", base), bulk.getAsJsonObject("cursor").get("query_id").getAsString());
    }

}
